### DISCUSSION TITLE
Handling multiple configurations

### DISCUSSION THREAD

**Discussion by katexochen** [golangci/golangci-lint#2432] [Q&A] [answered]
Hi, I'm am currently using different complex configurations, each defined in a separate yaml config file. Is there any best practice for handling multiple configurations? Moving multiple files between hosts and changing file path in the command line can be a lot of work. Ideally, I would like to define all configurations in the same yaml file and call them like 

```
golangci-lint run x
```
Thanks for helping!

**Comment by bombsimon** [comment, marked as answer]
There's no official or suggested way to do this as far as I know. If you actually have all the files on the machine I think just using the `--config` flag is what would be easiest. However it sounds like moving the files around is something you want to avoid and support for remote configuration has been requested in #2137. That issue contain some simple suggestions by using `make` and `curl` but that's not really a part of `golangci-lint`.

In #2241 there was an implementation to use different local files as `presets` but also a way to fetch presets remote with the same mechanism as used by `go modules`. However, the author of the PR closed it since no one did a proper review but personally I would be open to bring that back to life.

**Reply by katexochen** [reply]
I've took a look at those two PRs, and had the following thoughts:

#2137:
- A make file is a good solution to solve this problem on a project base. But I like how golangci-lint's config works in a global way. Maybe this could be solved in a similar way with make, my knowledge about make is somewhat limited.

#2241: 
- Seems to be a cool idea, but also has some technical complexity, and without the go module like part it wouldn't provide much benefit, since I would still have to manage multiple local files.

@bombsimon what do you think of my idea to define multiple configurations in a single file? This would be a simpler solution and would have minimal change to the UI. 

Just add a `name` field to the config file and separate different configurations with `---`:

```yml
name: my-check-1
run:
...

---
name: my-check-2
run:
...
```
And pass name via a new flag or preset:

```bash
golangci-lint --config-name my-check-1
# or via preset, using special character to separate from default presets
golangci-lint --preset @my-check-1 
```

**Reply by bombsimon** [reply]
Sorry for not getting back earlier. I have no strong feelings around your suggestion. I'm not sure I see the benefits of maintaining several documents in one file versus multiple files but I guess it boils down to how they're distributed. Personally I think it's a good idea to revision control your `golangci-lint` configuration(s) along with your code in the same repository so to me it doesn't make much difference to specify `--config-name` or just use `--config` which is already supported. I do however realize that people use different workflows and distribute our code and configuratiton differently so it makes more sense to some people compared to others. Generally though I prefer less code which is less maintanance.

**Reply by katexochen** [reply]
Thanks for your answer and your honest opinion. :)
